graphqa/util.py

```diff
@@ -0,0 +1,168 @@
+from collections import Counter
+import numpy as np
+import string
+import re
+import tensorflow as tf
+
+import logging
+logger = logging.getLogger(__name__)
+
+
+UNK = "<unk>"
+SOS = "<sos>"
+EOS = "<eos>"
+SPACE = "<space>"
+special_tokens = [UNK, SOS, EOS, SPACE]
+
+UNK_ID = special_tokens.index(UNK)
+SOS_ID = special_tokens.index(SOS)
+EOS_ID = special_tokens.index(EOS)
+
+CYPHER_PUNCTUATION = "()[]-=\"',.;:?"
+ENGLISH_PUNCTUATION = '!"#$%&()*+,-./:;=?@[\\]^_`{|}~'
+
+
+def load_vocab(args):
+    """
+    Load the tokens from the file constructed by "build_vocab" function in
+    build_data.py.
+    """
+    tokens = list()
+
+    with tf.gfile.GFile(args["vocab_path"]) as file:
+        for line in file.readlines():
+            tokens.append(line.replace("\n", ""))
+            if len(tokens) == args["vocab_size"]:
+                return tokens
+
+    return tokens
+
+
+def expand_unknown_vocab(line, vocab):
+    """
+    Treat the words that are in the "line" but not in the "vocab" as unknows,
+    and expand the characters in those words as individual words.
+    For example, the word "Spoon" is in "line" but not in "vocab", it will be
+    expanded as "<S> <p> <o> <o> <n>".
+    """
+    ts = set(line.split(' '))
+    unknowns = ts
+    unknowns -= set(vocab)
+    unknowns -= set([''])
+
+    for t in unknowns:
+        spaced = ''.join([f"<{c}> " for c in t])
+        line = line.replace(t, spaced)
+
+    return line
+
+
+def pretokenize_general(text):
+    """ Remove '\s*$' and replace ' ' by ' <space> '"""
+    text = re.sub(r'\s*$', '', text)
+    text = text.replace(" ", f" {SPACE} ")
+    return text
+
+
+def pretokenize_cypher(text):
+    """
+    First remove '\s*$' and replace ' ' by ' <space> '
+    Then add spaces before and after "()[]-=\"',.;:?"
+    Also treat spaces as tokens.
+    """
+    # In Cypher we want to tokenize punctuation as brackets are important
+    # therefore we treat spaces as a token as well
+    # so we can later reconstruct them
+
+    text = pretokenize_general(text)
+
+    for p in CYPHER_PUNCTUATION:
+        text = text.replace(p, f" {p} ")
+        # text = text.replace("  ", " ")
+    return text
+
+
+def pretokenize_english(text):
+    """
+    First remove '\s*$' and replace ' ' by ' <space> '
+    Then add spaces before and after '!"#$%&()*+,-./:;=?@[\\]^_`{|}~'
+    """
+    text = pretokenize_general(text)
+    for p in ENGLISH_PUNCTUATION:
+        text = text.replace(p, f" {p} ")
+
```

Comment:
`replace` will be extremely slow, a better version could be `translate`, e.g. 
```
table = str.maketrans({p: f" {p} " for p in ENGLISH_PUNCTUATION})
text.translate(table)
```

Also, RegExp could beat `translate` if you want to optimize it further.